plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMergeSink.java

```diff
@@ -355,20 +352,9 @@ private FileWriter createParquetFileWriter(String path, List<DeltaLakeColumnHand
         try {
             Closeable rollbackAction = () -> fileSystem.deleteFile(path);
 
-            List<Type> parquetTypes = dataColumns.stream()
-                    .map(column -> {
-                        Type type = column.getType();
-                        if (type instanceof TimestampWithTimeZoneType timestamp) {
-                            verify(timestamp.getPrecision() == 3, "Unsupported type: %s", type);
-                            return TIMESTAMP_MILLIS;
-                        }
-                        return type;
-                    })
-                    .collect(toImmutableList());
-
-            List<String> dataColumnNames = dataColumns.stream()
-                    .map(DeltaLakeColumnHandle::getName)
-                    .collect(toImmutableList());
+            List<String> dataColumnNames = dataColumns.stream().map(DeltaLakeColumnHandle::getPhysicalName).collect(toImmutableList());
+            List<Type> parquetTypes = dataColumns.stream().map(DeltaLakeColumnHandle::getSupportedType).collect(toImmutableList());
```

Comment:
Let's try to avoid iterating over the column list twice
```suggestion
ImmutableList.Builder<String> dataColumnNames = ImmutableList.builder();
ImmutableList.Builder<Type> parquetTypes = ImmutableList.builder();
for (DeltaLakeColumnHandle column : dataColumns) {
    dataColumnNames.add(..);
     parquetTypes.add(...);
}
```